Expose ExecutionPlan in prep for function calls

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21515 Add WeakIValue, use in tracer.
* #21514 clean up the TracingState API
* #21513 Collapse tracing_state.h into tracer.h
* #21325 Interpreter support for CallFunction/CallMethod
* **#21273 Expose ExecutionPlan in prep for function calls**
* #21272 Add flag to temporarily enable first class modules
* #21240 Reduce number of stack manipulation instructions in interpreter.
* #21185 Prepare interpreter for function calling

Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

Differential Revision: [D15600067](https://our.internmc.facebook.com/intern/diff/D15600067)

zdevito committed:
Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

zdevito committed:
Update on "Expose ExecutionPlan in prep for function calls"

Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

gh-metadata: pytorch pytorch 21273 gh/zdevito/47/head

Krovatkin commented on torch/csrc/jit/graph_executor_impl.h:
```diff
@@ -74,7 +52,44 @@ struct GraphExecutorImplBase {
 
   // entry point where execution begins
-  virtual void run(Stack& stack) = 0;
+  void run(Stack& stack);
```
i think we might need `virtual` `GraphExecutorImpl::run` eventually, but I guess, we can make it `virtual` when we need it

zdevito replied:
It would be possible. Keep in mind that run will only get called on entry from C++ into TorchScript. Function calls go directly through the getPlan pathway. So anything that has to happen for every function call will need to be expressed as part of getPlan.

Krovatkin reviewed (approved):
:shipit:

zdevito committed:
Update on "Expose ExecutionPlan in prep for function calls"

Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

gh-metadata: pytorch pytorch 21273 gh/zdevito/47/head

Commit:
Update on "Expose ExecutionPlan in prep for function calls"

Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

gh-metadata: pytorch pytorch 21273 gh/zdevito/47/head

Commit:
Update on "Expose ExecutionPlan in prep for function calls"

Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

gh-metadata: pytorch pytorch 21273 gh/zdevito/47/head

zdevito committed:
Update on "Expose ExecutionPlan in prep for function calls"

Expose ExecutionPlan in prep for function calls
Summary: This reorganizes GraphExecutor into a stage that looks up
the execution plan getPlanFor, and a stage that runs the plan. This
separation is necessary for first-class functions since the interpreter
will call getPlanFor to look up the Code for the function it will be
calling rather than recursively calling run on the graph executor.

Test Plan: test_jit.py

gh-metadata: pytorch pytorch 21273 gh/zdevito/47/head
